### ISSUE TITLE
Event publishing is failing with 500 in end-to-end test.

### ISSUE THREAD

**Issue by rakesh-garimella** [kyma-project/kyma#5196]
**Description**

End to end tests are failing because of error 500 on event publish.

**Expected result**

it should not fail

**Actual result**
it fails while publishing event.

**Troubleshooting**

Following are logs:
```
time="2019-08-05T06:39:03Z" level=info msg="Cleanup mode: yes"
time="2019-08-05T06:39:03Z" level=info msg="Step: 'Parallelized: Create namespace e2e-test, Create application e2e-test'"
time="2019-08-05T06:39:03Z" level=info msg="Step: 'Parallelized: Create mapping e2e-test, Deploy lambda e2e-test, Start test server, Connect application'"
time="2019-08-05T06:39:54Z" level=info msg="Step: 'Register test service'"
time="2019-08-05T06:39:55Z" level=info msg="Step: 'Create service instance'"
time="2019-08-05T06:40:03Z" level=info msg="Step: 'Create service binding'"
time="2019-08-05T06:40:03Z" level=info msg="Step: 'Create service binding usage'"
time="2019-08-05T06:40:03Z" level=info msg="Step: 'Create subscription'"
time="2019-08-05T06:40:03Z" level=info msg="Step: 'Retried: Send event, Check counter pod'"
time="2019-08-05T06:41:03Z" level=error msg="Error in 'Retried: Send event, Check counter pod': All attempts fail:\n#1: the counter pod is not updated: All attempts fail:\n#1: counter different then expected value: 0\n#2: counter different then expected value: 0\n#2: the counter pod is not updated: All attempts fail:\n#1: counter different then expected value: 0\n#2: counter different then expected value: 0\n#3: send event failed: 500\nrequest: &{POST https://gateway.gke-central-commit-6bfc58aa-5gtoqqf39r.a.build.kyma-project.io/e2e-test/v1/events HTTP/1.1 1 1 map[Content-Type:[application/json]] {0xc4203aec30} 0x656d70 170 [] false gateway.gke-central-commit-6bfc58aa-5gtoqqf39r.a.build.kyma-project.io map[] map[] <nil> map[]   <nil> <nil> <nil> <nil>}\nresponse: &{500 Internal Server Error 500 HTTP/1.1 1 1 map[Content-Length:[123] Content-Type:[application/json] Date:[Mon, 05 Aug 2019 06:40:06 GMT] Server:[istio-envoy] X-Envoy-Upstream-Service-Time:[96]] 0xc420404540 123 [] false false map[] 0xc4203b7d00 0xc42043a160}\n#4: send event failed: 500\nrequest: &{POST https://gateway.gke-central-commit-6bfc58aa-5gtoqqf39r.a.build.kyma-project.io/e2e-test/v1/events HTTP/1.1 1 1 map[Content-Type:[application/json]] {0xc42038e540} 0x656d70 170 [] false gateway.gke-central-commit-6bfc58aa-5gtoqqf39r.a.build.kyma-project.io map[] map[] <nil> map[]   <nil> <nil> <nil> <nil>}\nresponse: &{500 Internal Server Error 500 HTTP/1.1 1 1 map[Content-Length:[123] Content-Type:[application/json] Date:[Mon, 05 Aug 2019 06:40:06 GMT] Server:[istio-envoy] X-Envoy-Upstream-Service-Time:[25]] 0xc4204aada0 123 [] false false map[] 0xc420170100 0xc420102630}\n#5: the counter pod is not updated: All attempts fail:\n#1: counter different then expected value: 0\n#2: counter different then expected value: 0\n#6: the counter pod is not updated: All attempts fail:\n#1: counter different then expected value: 0\n#2: counter different then expected value: 0\n#7: the counter pod is not updated: All attempts fail:\n#1: counter different then expected value: 0\n#2: counter different then expected value: 0\n#8: the counter pod is not updated: All attempts fail:\n#1: counter different then expected value: 0\n#2: counter different then expected value: 0\n#9: the counter pod is not updated: All attempts fail:\n#1: counter different then expected value: 0\n#2: counter different then expected value: 0\n#10: the counter pod is not updated: All attempts fail:\n#1: counter different then expected value: 0\n#2: counter different then expected value: 0"
time="2019-08-05T06:41:03Z" level=info msg="Cleanup: 'Retried: Send event, Check counter pod'"
time="2019-08-05T06:41:03Z" level=info msg="Cleanup: 'Create subscription'"
time="2019-08-05T06:41:03Z" level=info msg="Cleanup: 'Create service binding usage'"
time="2019-08-05T06:41:03Z" level=info msg="Cleanup: 'Create service binding'"
time="2019-08-05T06:41:04Z" level=info msg="Cleanup: 'Create service instance'"
time="2019-08-05T06:41:07Z" level=info msg="Cleanup: 'Register test service'"
time="2019-08-05T06:41:07Z" level=info msg="Cleanup: 'Parallelized: Create mapping e2e-test, Deploy lambda e2e-test, Start test server, Connect application'"
time="2019-08-05T06:41:58Z" level=info msg="Cleanup: 'Parallelized: Create namespace e2e-test, Create application e2e-test'"
[0m[1mEnd of testing 'oct-tp-testsuite-all-2019-08-05-06-32-core-test-external-solution-0'
```

**Comment by montaro** [comment]
After trying to reproduce, my guess is:
Running the E2E tests to reproduce:
```
❯ go run ./cmd/runner e2e --domain 34.77.147.40.xip.io --skipSSLVerify
INFO[0000] Cleanup mode: yes
INFO[0000] Step: 'Parallelized: Create namespace e2e-test, Create application e2e-test'
INFO[0000] Step: 'Parallelized: Create mapping e2e-test, Deploy lambda e2e-test, Start test server, Connect application'
INFO[0032] Step: 'Register test service'
INFO[0033] Step: 'Create service instance'
INFO[0044] Step: 'Create service binding'
INFO[0044] Step: 'Create service binding usage'
INFO[0044] Step: 'Create subscription'
INFO[0044] Step: 'Retried: Send event, Check counter pod'
INFO[0065] Cleanup: 'Retried: Send event, Check counter pod'
INFO[0065] Cleanup: 'Create subscription'
INFO[0065] Cleanup: 'Create service binding usage'
INFO[0065] Cleanup: 'Create service binding'
INFO[0067] Cleanup: 'Create service instance'
INFO[0070] Cleanup: 'Register test service'
INFO[0071] Cleanup: 'Parallelized: Create mapping e2e-test, Deploy lambda e2e-test, Start test server, Connect application'
INFO[0122] Cleanup: 'Parallelized: Create namespace e2e-test, Create application e2e-test'
INFO[0135]/Users/D073502/workspace/sap-cx/go/src/github.com/kyma-project/kyma/tests/end-to-end/external-solution-integration/cmd/runner/main.go:57 main.main() Successfully Finished the e2e test!!
```
I think this means that sending the message failed in the very first time and succeeded later.
Checking the logs in the Event-Bus-Publish app:
```
event-bus-event-publish-service-948f6466c-ks55z event-publish-service 2019/08/08 14:01:14 publish to the knative channel: 'kgdqt7qn3c5frleefibvfdlzv2jezs4gj' namespace: 'kyma-system' status: 'ignored' reason: 'Event was ignored as there are no subscriptions or consumers configured for this event'
event-bus-event-publish-service-948f6466c-ks55z event-publish-service 2019/08/08 14:01:15 ERROR: SendMessage(): 500 Internal Server Error
event-bus-event-publish-service-948f6466c-ks55z event-publish-service 2019/08/08 14:01:15 failed to send message to the knative channel 'kgdqt7qn3c5frleefibvfdlzv2jezs4gj' in namespace 'kyma-system'
event-bus-event-publish-service-948f6466c-ks55z event-publish-service 2019/08/08 14:01:16 publish to the knative channel: 'kgdqt7qn3c5frleefibvfdlzv2jezs4gj' namespace: 'kyma-system' status: 'published' reason: 'Message successfully published to the channel'
```
I believe this enforces the assumption that the first one failed then it succeeded to deliver.
Checking the NATS-Dispatcher logs:
```
natss-dispatcher-9cd9cd9cb-knxlv dispatcher {"level":"info","ts":1565272874.893683,"caller":"channel/reconcile.go:56","msg":"Reconcile: ","request":"kyma-system/kgdqt7qn3c5frleefibvfdlzv2jezs4gj"}
natss-dispatcher-9cd9cd9cb-knxlv dispatcher {"level":"error","ts":1565272874.974151,"caller":"channel/reconcile.go:90","msg":"Error updating Channel Status","error":"Operation cannot be fulfilled on channels.eventing.knative.dev \"kgdqt7qn3c5frleefibvfdlzv2jezs4gj\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"github.com/knative/eventing/contrib/natss/pkg/dispatcher/channel.(*reconciler).Reconcile\n\t/home/prow/go/src/github.com/knative/eventing/contrib/natss/pkg/dispatcher/channel/reconcile.go:90\ngithub.com/knative/eventing/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/home/prow/go/src/github.com/knative/eventing/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215\ngithub.com/knative/eventing/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/home/prow/go/src/github.com/knative/eventing/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158\ngithub.com/knative/eventing/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/home/prow/go/src/github.com/knative/eventing/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133\ngithub.com/knative/eventing/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/home/prow/go/src/github.com/knative/eventing/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134\ngithub.com/knative/eventing/vendor/k8s.io/apimachinery/pkg/util/wait.Until\n\t/home/prow/go/src/github.com/knative/eventing/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88"}
natss-dispatcher-9cd9cd9cb-knxlv dispatcher {"level":"info","ts":1565272875.5548258,"caller":"provisioners/message_receiver.go:140","msg":"Received request for kgdqt7qn3c5frleefibvfdlzv2jezs4gj-channel-b92xq.kyma-system.svc.cluster.local"}
natss-dispatcher-9cd9cd9cb-knxlv dispatcher {"level":"info","ts":1565272875.5548997,"caller":"provisioners/message_receiver.go:143","msg":"Could not extract channel","error":"Invalid HostName:\"kgdqt7qn3c5frleefibvfdlzv2jezs4gj-channel-b92xq.kyma-system.svc.cluster.local\". HostName not found in any of the watched natss channels"}
natss-dispatcher-9cd9cd9cb-knxlv dispatcher {"level":"info","ts":1565272875.9745457,"caller":"channel/reconcile.go:56","msg":"Reconcile: ","request":"kyma-system/kgdqt7qn3c5frleefibvfdlzv2jezs4gj"}
natss-dispatcher-9cd9cd9cb-knxlv dispatcher {"level":"info","ts":1565272875.986444,"caller":"channel/reconcile.go:56","msg":"Reconcile: ","request":"kyma-system/kgdqt7qn3c5frleefibvfdlzv2jezs4gj"}
natss-dispatcher-9cd9cd9cb-knxlv dispatcher {"level":"info","ts":1565272875.9865239,"caller":"dispatcher/dispatcher.go:243","msg":"Subscribe to channel:","channel":{"Namespace":"kyma-system","Name":"kgdqt7qn3c5frleefibvfdlzv2jezs4gj"},"subscription":{"UID":"fc6d3f47-b9e4-11e9-b52e-42010a8402ab","SubscriberURI":"http://e2e-test.e2e-test:8080","ReplyURI":""}}
...
```
I don't have a clear explanation why this is happening and my suggestion is to wait untill we update the k8s cluster because it might be a bug: https://github.com/kubernetes-sigs/controller-runtime/issues/403

**Comment by sayanh** [comment]
I'm curious when the publisher logged `500`, were the (knative)channel and (knative)subscription ready? My feeling is they were not. Hence, IMO there is a direct correlation to this [issue](https://github.com/kyma-project/kyma/issues/4919)

**Comment by montaro** [comment]
So the result of the investigation is that:
1- The message gets ignored in the very first trials because of timing as explained earlier, no workaround needs to be done at the moment because the event publish will succeed eventually (No sleep needs to be added to the test). The fix should be: Change the subscription controller to mark the kyma-subscription as ready only if the knative-subscription is ready (implies the channel is ready)
2- The 500 is related to the old `mTLS` problem which was closed as @sayanh was not able to reproduce it back then.
The ticket will be closed for now since we are moving to a Broker-Trigger Event Bus soon!

**Comment by marcobebway** [comment]
Closing this ticket and marking it as won't fix.

**Comment by sjanota** [comment]
The error still happens. Example runs:
https://status.build.kyma-project.io/view/gcs/kyma-prow-logs/logs/post-master-kyma-gke-integration/1164166641689300994

**Comment by sayanh** [comment]
So this is what happens here:

After Kyma subscription is created, subscription controller goes on to create a Knative channel. Once this is done, natss-dispatcher does reconcile and creates a subscription is Natss server. It also holds the channel information in an internal map `hostToChannelMap` which it references when a publish is done. Now when a publish is done right after creating a Kyma subscription, the Natss dispatcher hasn't reconciled yet. Hence the internal map `hostToChannelMap` does not have the channel information hence it fails with the following logs:

```
{"level":"info","ts":1568303550.7187064,"caller":"provisioners/message_receiver.go:140","msg":"Received request for kgdqt7qn3c5frleefibvfdlzv2jezs4gj-channel-8qfz2.kyma-system.svc.cluster.local"}
{"level":"info","ts":1568303550.718754,"caller":"provisioners/message_receiver.go:143","msg":"Could not extract channel","error":"Invalid HostName:\"kgdqt7qn3c5frleefibvfdlzv2jezs4gj-channel-8qfz2.kyma-system.svc.cluster.local\". HostName not found in any of the watched natss channels"}
{"level":"info","ts":1568303550.942892,"caller":"channel/reconcile.go:56","msg":"Reconcile: ","request":"kyma-system/kgdqt7qn3c5frleefibvfdlzv2jezs4gj"}
{"level":"info","ts":1568303551.0526907,"caller":"channel/reconcile.go:56","msg":"Reconcile: ","request":"kyma-system/kgdqt7qn3c5frleefibvfdlzv2jezs4gj"}
{"level":"info","ts":1568303551.0528836,"caller":"dispatcher/dispatcher.go:243","msg":"Subscribe to channel:","channel":{"Namespace":"kyma-system","Name":"kgdqt7qn3c5frleefibvfdlzv2jezs4gj"},"subscription":{"UID":"537a20d9-d575-11e9-9c61-42010a9c00da","SubscriberURI":"http://e2e-test.e2e-test:8080","ReplyURI":""}}
{"level":"info","ts":1568303551.0708466,"caller":"dispatcher/dispatcher.go:280","msg":"NATSS Subscription created: &{RWMutex:{w:{state:0 sema:0} writerSem:0 readerSem:0 readerCount:0 readerWait:0} sc:0xc0000f6540 subject:kgdqt7qn3c5frleefibvfdlzv2jezs4gj.kyma-system qgroup: inbox:_INBOX.lBa4B3AMdGc2MACj5Sc2hv ackInbox:_INBOX.URwnk7PiBtcNm4nmv7SdMC inboxSub:0xc000211980 opts:{DurableName:537a20d9-d575-11e9-9c61-42010a9c00da MaxInflight:1024 AckWait:60000000000 StartAt:0 StartSequence:0 StartTime:{wall:0 ext:0 loc:<nil>} ManualAcks:true} cb:0x10f77d0}"}
{"level":"info","ts":1568303551.0711122,"logger":"fallback","caller":"dispatcher/dispatcher.go:321","msg":"hostToChannelMap updated successfully."}
```

The logs also show that afterwards, the natss-dispatcher reconciles meaning a subscription gets created in Natss server and the internal map is updated.
